### PR TITLE
Protect against LTO optimizer

### PR DESCRIPTION
**Pull request by rhc54**
We currently do not support the LTO optimizer
as it is incompatible with our plugin component
architecture. So detect it has been specified
in configure and error out with an explanation.

### CODE REVIEW

File: config/pmix_check_cflags.m4 [hunk 1]
```diff
@@ -40,3 +40,14 @@ AC_MSG_CHECKING(if $CC supports ([$1]))
                 AC_MSG_RESULT([yes])
             fi
 ])
+
+AC_DEFUN([_PMIX_CHECK_LTO_FLAG], [
+    chkflg=`echo $1 | grep -- -flto`
+    if test "$chkflg"; then
```
Comment (jsquyres) [inline config/pmix_check_cflags.m4:46]:
```suggestion
    if test -n "$chkflg"; then
```

I can't remember offhand if `test "$foo"` is the same as `test -n "$foo"`.

I'm 100% sure that `-n` means "check for a non-empty string".  I don't remember offhand if having no test operator means "check for a true-ish value" or "check for a non-empty string".

File: config/pmix.m4 [hunk 2]
```diff
@@ -890,6 +890,17 @@ AC_DEFUN([PMIX_SETUP_CORE],[
     CPP_INCLUDES="$(echo $cpp_includes | $SED 's/[[^ \]]* */'"$pmix_cc_iquote"'&/g')"
     CPPFLAGS="$CPP_INCLUDES -I$PMIX_top_srcdir/include $CPPFLAGS"
 
+
+    # We do not currently support the "lto" optimizer as it
+    # aggregates all the headers from our plugins, resulting
+    # in a configuration that generates warnings/errors when
+    # passed through their optimizer faze. We therefore check
```
Comment (jsquyres) [inline config/pmix.m4:897]:
```suggestion
    # passed through their optimizer phase. We therefore check
```

File: config/pmix.m4 [hunk 3]
```diff
@@ -890,6 +890,17 @@ AC_DEFUN([PMIX_SETUP_CORE],[
...
+    # for the flag, and if found, output a message explaining
+    # the situation and aborting configure
+    _PMIX_CHECK_LTO_FLAG($CPPFLAGS, CPPFLAGS)
+    _PMIX_CHECK_LTO_FLAG($CFLAGS, CFLAGS)
```
Comment (jsquyres) [inline config/pmix.m4:901]: Might want to check $LDFLAGS and $LIBS, too.

I *think* the most appropriate location for `-flto` is in `$LDFLAGS` (it's a linker flag, after all), but people could choose to put it in any weird location (including $CFLAGS, $CPPFLAGS, $LIBS, etc.).

File: config/pmix_check_cflags.m4 [hunk 4]
```diff
@@ -40,3 +40,14 @@ AC_MSG_CHECKING(if $CC supports ([$1]))
...
+        AC_MSG_WARN([Configure has detected the presence of the -flto])
+        AC_MSG_WARN([compiler directive in $2. PMIx does not currently])
+        AC_MSG_WARN([support this optimizer as it conflicts with the])
```
Comment (jsquyres) [inline config/pmix_check_cflags.m4:49]:
```suggestion
        AC_MSG_WARN([support this flag as it conflicts with the])
```

File: config/pmix_check_cflags.m4 (hunk 1 above)
Comment (rhc54) [reply config/pmix_check_cflags.m4:46]: It is the same, but no harm in adding the `-n` to make it clear.

**Review by jsquyres — APPROVED** [review]

File: config/pmix.m4 (hunk 3 above)
Comment (eli-schwartz) [reply config/pmix.m4:901]: It's actually "required" to be specified both at compile time and link time.
- If you pass it at link time only, it doesn't find any LTO btyecode to perform LTO on, and it's way too late to convert source code into bytecode.
- If you pass it at compile time only, the compiler *may* (probably does! I know gcc does, but I can't remember what clang does) choose to infer from the presence of bytecode in the object files, that it can do LTO. Object files can double-compile and produce both object code and bytecode, in which case you can choose to disable it at link time (-fno-lto) and it simply ignores the bytecode...

Basically the tl;dr here is that checking in CFLAGS is sufficient but checking in both won't hurt.
